packages/eslint-parser/src/common/espree.ts

```diff
@@ -0,0 +1,141 @@
+import path from 'node:path'
+import process from 'node:process'
+
+// @ts-expect-error -- ignore
+import * as dependencyEspree from 'espree'
+import { lt, lte } from 'semver'
+import type { VineESLintParserOptions } from '../types'
+import { getLinterRequire } from './linter-require'
+
+import { createRequire } from './create-require'
+import type { BasicParserObject } from './parser-object'
+
+type Espree = BasicParserObject & {
+  latestEcmaVersion?: number
+  version: string
+}
+let espreeCache: Espree | null = null
+
+/**
+ * Gets the espree that the given ecmaVersion can parse.
+ */
+export function getEspreeFromEcmaVersion(
+  ecmaVersion: VineESLintParserOptions['ecmaVersion'],
+): Espree {
+  const linterEspree = getEspreeFromLinter()
+  if (ecmaVersion == null) {
+    return linterEspree
+  }
+  if (ecmaVersion === 'latest') {
+    return getNewestEspree()
+  }
+  if (
+    normalizeEcmaVersion(ecmaVersion) <= getLatestEcmaVersion(linterEspree)
+  ) {
+    return linterEspree
+  }
+  const userEspree = getEspreeFromUser()
+  if (normalizeEcmaVersion(ecmaVersion) <= getLatestEcmaVersion(userEspree)) {
+    return userEspree
+  }
+  return linterEspree
+}
+
+/**
+ * Load `espree` from the user dir.
+ */
+export function getEspreeFromUser(): Espree {
+  try {
+    const cwd = process.cwd()
+    const relativeTo = path.join(cwd, '__placeholder__.js')
+    return createRequire(relativeTo)('espree')
+  }
+  catch {
+    return getEspreeFromLinter()
+  }
+}
+
+/**
+ * Load `espree` from the loaded ESLint.
+ * If the loaded ESLint was not found, just returns `require("espree")`.
+ */
+export function getEspreeFromLinter(): Espree {
+  if (!espreeCache) {
+    espreeCache = getLinterRequire()?.('espree')
+    if (!espreeCache) {
```

Comment:
refatctor:
- espreeCache = getLinterRequire()?.('espree') || dependencyEspree

Reply:
Simon you could use this `Add a suggestion` button to create a review suggestion diff block, and we can apply that diff changes directly on Github web pages. 🙏